### PR TITLE
Support retrieving widechar data in parts

### PR DESCRIPTION
**Pull request by staticlibs**
This change implements support for returning variable-length data to client with multiple subsequent calls to `SQLGetData`. The ODBC spec logic for buffer/write/return sizing is relatively straightforward, but there is a number of corner cases. Also the logic for char, widechar and binary data differs slightly (different null-termination handling). This change unifies this logic for all variable-length types in one place.

Testing: new test is added that covers retrieving data by parts for char, widechar and binary data. PyODBC test suite is extended to check the retrieval of overly long strings and also enabled on CI on Windows (to check the retrieval through the Windows Driver Manager).

Fixes: #121

### CODE REVIEW

File: src/statement/statement_functions.cpp [hunk 1]
```diff
@@ -194,50 +194,94 @@ static bool CastTimestampValue(duckdb::OdbcHandleStmt *hstmt, const duckdb::Valu
...
+//  - SQL_CHAR: 1 byte
+//  - SQL_BINARY: None
+template <typename CHAR_TYPE>
+static SQLRETURN GetVariableValue(SQLUSMALLINT col_idx, duckdb::OdbcHandleStmt *hstmt, SQLPOINTER target_value_ptr,
+                                  SQLLEN buffer_length, SQLLEN *str_len_or_ind_ptr, CHAR_TYPE *val_buf,
+                                  std::size_t val_buf_len_bytes, bool null_terminate = true) {
+
+	// Clear up remembered length if the column has changed.
+	hstmt->odbc_fetcher->SetLastFetchedVariableVal(static_cast<duckdb::row_t>(col_idx));
+
```
Comment (maiadegraaf) [inline src/statement/statement_functions.cpp:215]: Nit: Do you mean something like`Reset stored length if the column has changed`?

Not entirely sure what's happening here. Does this function reset only if the `col_idx` has changed?

File: src/statement/statement_functions.cpp [hunk 2]
```diff
@@ -194,50 +194,94 @@ static bool CastTimestampValue(duckdb::OdbcHandleStmt *hstmt, const duckdb::Valu
...
 
-	if (str_len_or_ind_ptr) {
-		*str_len_or_ind_ptr = out_len;
+	// Write the length of the data after conversion and before truncation due to BufferLength.
+	// Null-termination bytes are excluded.
+	if (str_len_or_ind_ptr != nullptr) {
+		*str_len_or_ind_ptr = remaining_len;
 	}
 
+	// Save the lenght of data already writtent for future calls
```
Comment (maiadegraaf) [inline src/statement/statement_functions.cpp:282]: Nit:
```suggestion
	// Save the length of data already written for future calls
```

File: src/statement/statement_functions.cpp [hunk 3]
```diff
@@ -377,51 +421,24 @@ SQLRETURN duckdb::GetDataStmtResult(OdbcHandleStmt *hstmt, SQLUSMALLINT col_or_p
...
+		std::string val_str = val.GetValue<std::string>();
+		// We need to convert the result to UTF-16 to get its length (in bytes) even if we are not going to return it
+		auto utf16_vec = duckdb::widechar::utf8_to_utf16_lenient(reinterpret_cast<const SQLCHAR *>(val_str.c_str()),
+		                                                         val_str.length());
+		return GetVariableValue(col_or_param_num, hstmt, target_value_ptr, buffer_length, str_len_or_ind_ptr,
+		                        utf16_vec.data(), utf16_vec.size() * sizeof(SQLWCHAR));
 	}
 	// case SQL_C_VARBOOKMARK: // same ODBC type (\\TODO we don't support bookmark types)
 	case SQL_C_BINARY: {
 		// threating binary values as BLOB type
```
Comment (maiadegraaf) [inline src/statement/statement_functions.cpp:433]: Nit:
``` suggestion
		// treating binary values as BLOB type
```

**Review by maiadegraaf — COMMENTED** [review]
Hi @staticlibs! Thanks for the PR, looks like a far more comprehensive and complete approach to `SQLGetData`. The test coverage for handling large data is way beter, and well the pyodbc testing is improved, but I think this still can be further expanded in the future :)

Just left a few nits.

File: src/statement/statement_functions.cpp (hunk 3 above)
Comment (staticlibs) [reply src/statement/statement_functions.cpp:433]: Fixed.

**Comment by staticlibs** [comment]
Thanks for the review! Updated the commit with typos fixed.

**Comment by maiadegraaf** [comment]
Great, feel free to merge once tests have passed :)
